msal_extensions/osx.py

```diff
@@ -0,0 +1,315 @@
+    Keychain type."""
+    def __init__(self, exit_status):
+        super(KeychainError, self).__init__()
+        self.exit_status = exit_status
+        # TODO: pylint: disable=fixme
+        #  use SecCopyErrorMessageString to fetch the appropriate message here.
+        self.message = \
+            '{} ' \
+            'see https://opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/MacErrors.h'\
+                .format(self.exit_status)
+
+
+class KeychainAccessDeniedError(KeychainError):
+    """The exceptions that's raised when a Keychain is set to be queried, but the current process
+    has not been give permission to access it, and a query comes anyway."""
+    EXIT_STATUS = -128
+
+    def __init__(self):
+        super(KeychainAccessDeniedError, self).__init__(KeychainAccessDeniedError.EXIT_STATUS)
+
+
+class NoSuchKeychainError(KeychainError):
+    """The exception that's raised when a Keychain is set to be queried, but it is non-existent, and
+    a query comes anyway."""
+    EXIT_STATUS = -25294
+
+    def __init__(self, name):
+        super(NoSuchKeychainError, self).__init__(NoSuchKeychainError.EXIT_STATUS)
+        self.name = name
+
+
+class NoDefaultKeychainError(KeychainError):
+    """The exception that's raised when no Keychain is set to be queried, but a query comes
+    anyway.
+    """
+    EXIT_STATUS = -25307
+
+    def __init__(self):
+        super(NoDefaultKeychainError, self).__init__(NoDefaultKeychainError.EXIT_STATUS)
+
+
+class KeychainItemNotFoundError(KeychainError):
+    """The exception that's raised when a non-exist Keychain entry is requested."""
+    EXIT_STATUS = -25300
+
+    def __init__(self, service_name, account_name):
+        super(KeychainItemNotFoundError, self).__init__(KeychainItemNotFoundError.EXIT_STATUS)
+        self.service_name = service_name
+        self.account_name = account_name
+
+
+def _construct_error(exit_status, **kwargs):
+    if exit_status == KeychainAccessDeniedError.EXIT_STATUS:
+        return KeychainAccessDeniedError()
+    if exit_status == NoSuchKeychainError.EXIT_STATUS:
+        return NoSuchKeychainError(**kwargs)
+    if exit_status == NoDefaultKeychainError.EXIT_STATUS:
+        return NoDefaultKeychainError()
+    if exit_status == KeychainItemNotFoundError.EXIT_STATUS:
+        return KeychainItemNotFoundError(**kwargs)
+    return KeychainError(exit_status)
+
+
+def _get_native_location(name):
+    # type: (str) -> str
+    """
+    Fetches the location of a native MacOS library.
+    :param name: The name of the library to be loaded.
+    :return: The location of the library on a MacOS filesystem.
+    """
+    return '/System/Library/Frameworks/{0}.framework/{0}'.format(name)
+
+
+# Load native MacOS libraries
+try:
+    _SECURITY = _ctypes.CDLL(_get_native_location('Security'))
+    _CORE = _ctypes.CDLL(_get_native_location('CoreFoundation'))
+except OSError:
+    _SECURITY = None
```

Comment:
If I understand this correctly, when we "recover" from OSError and assign `_SECURITY` as `None`, the code below `SOMETHING = _SECURITY.SomeThing` would still blow up. So, shall we just do a logging here or - my personal favorite - not catching this OSError exception at all?

Reply:
I'm guessing I did this before I guarded my import statements by platform. 😝

Reply:
Fixed in 469da0073da2b480ec5ce9a2e78165cfa0b96cee